README.rst
==================================================================
Polyvers: Bump sub-project versions in Git monorepos independently
==================================================================

.. _opening-start:
.. image:: https://img.shields.io/pypi/v/polyvers.svg
    :alt: Deployed in PyPi?
    :target: https://pypi.org/pypi/polyvers

.. image:: https://img.shields.io/travis/JRCSTU/polyvers.svg
    :alt: TravisCI (linux) build ok?
    :target: https://travis-ci.org/JRCSTU/polyvers

.. image:: https://ci.appveyor.com/api/projects/status/lyyjtmit5ti7tg1n?svg=true
    :alt: Apveyor (Windows) build?
    :scale: 100%
    :target: https://ci.appveyor.com/project/ankostis/polyvers

.. image:: https://img.shields.io/coveralls/github/JRCSTU/polyvers.svg
    :alt: Test-case coverage report
    :scale: 100%
    :target: https://coveralls.io/github/JRCSTU/polyvers?branch=master&service=github

.. image:: https://readthedocs.org/projects/polyvers/badge/?version=latest
    :target: https://polyvers.readthedocs.io/en/latest/?badge=latest
    :alt: Auto-generated documentation status

.. image:: https://pyup.io/repos/github/JRCSTU/polyvers/shield.svg
    :target: https://pyup.io/repos/github/JRCSTU/polyvers/
    :alt: Dependencies needing updates?

.. image:: https://api.codacy.com/project/badge/Grade/11b2545fd0264f1cab4c862998833503
    :target: https://www.codacy.com/app/ankostis/polyvers_jrc
    :alt: Code quality metric

:version:       0.1.1a1
:updated:       2018-08-07T13:30:07.601687
:Documentation: https://polyvers.readthedocs.io
:repository:    https://github.com/JRCSTU/polyvers
:pypi-repo:     https://pypi.org/project/polyvers/, https://pypi.org/project/polyversion/
:keywords:      version-management, configuration-management, versioning, git, monorepos,
                tool, library
:copyright:     2018 JRC.C4(STU), European Commission (`JRC <https://ec.europa.eu/jrc/>`_)
:license:       `EUPL 1.2 <https://joinup.ec.europa.eu/software/page/eupl>`_

A Python 3.6+ command-line tool to manage `PEP-440 version-ids
<https://www.python.org/dev/peps/pep-0440/>`_ of dependent sub-projects
hosted in a *Git* :term:`monorepo`\s, independently.

The key features are:

- :term:`setuptools integration`,
- x2 :term:`repo scheme`\s (:term:`monorepo`, :term:`mono-project`),
- configurable :term:`version scheme`,
- *leaf* :term:`release scheme`,
- intuitive :term:`version-bump algebra` (TODO),
- configurable :term:`engravings`.

The *leaf version scheme* feature departs from the logic of :ref:`similar-tools`.
Specifically, when bumping the version of sub-project(s), this tool
adds **+2 tags and +1 commits**:

  - one :term:`Version tag` in-trunk like ``foo-proj-v0.1.0``,
  - and another :term:`Release tag` on a new :term:`out-of-trunk commit` (leaf)
    like ``foo-proj-r0.1.0`` (the new version-ids are :term:`engrave`\d only
    in this release-commit):

    .. figure:: _static/leaf_commits.png
        :align: center
        :alt: Leaf-commits & version/release tags for the two repo's sub-projects

        Leaf-commits & version/release-tags for the two sub-project's,
        as shown in this repo's git history.

.. Note::
  The reason for this feature is to allow exchange code across branches (for the
  different sub-projects) without :term:`engravings` getting in your way as
  merge-conflicts.

Additional capabilities and utilities:

- It is still possible to use plain **version tags (vtags)** like ``v0.1.0``,
  assuming you have a single project (called hereinafter a :term:`mono-project`)

- A separate Python 2.7+ **polyversion** project, which contains API to extract
  sub-project's version from past tags (provided as a separate subproject
  so client programs do not get ``polyvers`` commands transitive dependencies).
  The library functions as a :term:`setuptools  plugin`.

.. _opening-end:

.. contents:: Table of Contents
   :backlinks: top
   :depth: 4


.. _usage:

Tutorial
========
Install the tool
----------------
And you get the ``polyvers`` command:

.. code-block:: console

    $ pip install polyvers
    ...
    $ polyvers --version
    0.0.0
    $ polyvers --help
    ...

    $ polyvers status
    polyvers: Neither `setup.py` nor `.polyvers(.json|.py|.salt)` config-files found!

.. Note::
    Actually two projects are installed:

    - **polyvers** cmd-line tool, for developing python :term:`monorepo`\s,
    - **polyversion**: the base python library used by projects developed
      with *polyvers* tool, so that their sources can discover their subproject-version
      on runtime from Git.


Prepare project
---------------
Assuming our :term:`monorepo` project ``/monorepo.git/`` contains two sub-projects,
then you need enter the following configurations into your build files::

    /monorepo.git/
        +--setup.py               # see below for contents
        +--mainprog/__init__.py
        |                         from polyversion import polyversion, polytime
        |                         __version__ = polyversion()
        |                         __updated__ = polytime()
        |                         ...
        |
        +--core-lib/
            +--setup.py:          # like above
            +--core/__init__.py   # like above
            +--...

.. Tip::
    You may see different sample approaches for your setup-files by looking
    into both `polyvers` & `polyversion` subprojects of this repo (because
    they eat their own dog food).

The `polyversion` library function as a *setuptools* "plugin", and
adds a new ``setup()`` keyword ``polyversion = (bool | dict)``
(see :func:`polyversion.init_plugin_kw` for its content), which you can use it
like this:

.. code-block:: python

    from setuptools import setup

    setup(
        project='myname',
        version=''              # omit (or None) to abort if cannot auto-version
        polyversion={           # dict or bool
            'version_scheme: 'mono-project',
            ...  # See `polyversion.init_plugin_kw()` for more keys.
        },
        setup_requires=[..., 'polyversion'],
        ...
    )

.. Hint::
    The ``setup_requires=['polyvers']`` keyword  (only available with *setuptools*,
    and not *distutils*), enables the new ``polyversion=`` setup-keyword.

Alternatively, a subproject may use :pep:`0518` to pre-install `polyversion`
library *before* pip-installing or launching ``setup.py`` script.
To do that, add the ``pyproject.toml`` file below next to your `setup` script::

    [build-system]
    requires = ["setuptools", "wheel", "polyversion"]

and then you can simply import ``polyversion`` from your ``setup.py``:

.. code-block:: python

    from setuptools import setup
    from polyversion import polyversion

    setup(
        project='myname',
        version=polyversion(mono_project=True)  # version implied empty string.

.. Attention::
    To properly install a :pep:`0518` project you need ``pip-v10+`` version.


Initialize `polyvers`
---------------------
...we let the tool auto-discover the mapping of *project folders ↔ project-names*
and create a `traitlets configuration YAML-file <https://traitlets.readthedocs.io>`_
named as  ``/monorepo.git/.polyvers.py``:

.. code-block:: console

    $ cd monorepo.git

    $ polyvers init --monorepo
    Created new config-file '.polyvers.yaml'.

    $ cat .polyvers.yaml
    ...
    PolyversCmd:
      projects:
      - pname: mainprog     # name extracted from `setup.py`.
        basepath: .         # path discovered by the location of `setup.py`
      - pname: core
        basepath: core-lib
    ...

    $ git add .polyvers.yaml
    $ git commit -m 'add polyvers config-gile'

And now we can use the ``polyvers`` command to inspect the versions of all
sub-projects:

.. code-block:: console

    $ polyvers status
    - mainprog
    - core

Indeed there are no tags in in git-history for the tool to derive and display
project-versions, so only project-names are shown.  With ``--all`` option
more gets displayed:

.. code-block:: console

    $ polyvers status -a
    - pname: mainprog
      basepath: .
      gitver:
      history: []
    - pname: core
      basepath: core-lib
      gitver:
      history: []

..where ``gitver`` would be the result of ``git-describe``.


Bump versions
-------------
We can now use tool to set the same version to all sub-projects:

.. code-block:: console

    $ polyvers bump 0.0.0 -f noengraves   # all projects implied, if no project-name given
    00:52:06       |WARNI|polyvers.bumpcmd.BumpCmd|Ignored 1 errors while checking if at least one version-engraving happened:
      ignored (--force=noengraves): CmdException: No version-engravings happened, bump aborted.
    00:52:07       |NOTIC|polyvers.bumpcmd.BumpCmd|Bumped projects: mainprog-0.0.0 --> 0.0.0, core-0.0.0 --> 0.0.0

The ``--force=noengraves`` disables a safety check that requires at least one
file modification for :term:`engrave`\ing the current version in the leaf "Release" commit
(see next step).

.. code-block:: console

    $ polyvers status
    - mainprog-v0.0.0
    - core-v0.0.0

    $ git lg    # Ok, augmented `lg` output a bit here...HEAD --> UPPER branch.
    COMMITS BRANCH TAGS                 REMARKS
    ======= ====== ==================== ========================================
         O  latest mainprog-r0.0.0      - x2 tags on "Release" leaf-commit
        /          core-r0.0.0            outside-of-trunk (not in HEAD).
       O    MASTER mainprog-v0.0.0      - x2 tags on "Version" commit
       |           core-v0.0.0            for bumping both projects to v0.0.0
       O                                - Previous commit, before version bump.

   .. Hint::
      Note the difference between ``ABC-v0.0.0`` vs ``ABC-r0.0.0`` tags.

   In the source code, it's only the "release" commit that has :term:`engrave`\d* version-ids:

   .. code-block:: console

    $ cat mainprog/mainprog/__init__.py    # Untouched!
    import polyvers

    __title__     = "mainprog"
    __version__ = polyvers.version('mainprog')
    ...

    $ git checkout  latest
    $ cat mainprog/mainprog/__init__.py
    import polyvers

    __title__     = "mainprog"
    __version__ = '0.0.0'
    ...

    $ git checkout  -  # to return to master.


Engrave version in the sources
------------------------------
Usually programs report their version somehow when run, e.g. with ```cmd --version``.
With *polyvers* we can derive the latest from the tags created in the previous step,
using a code like this, usually in the file ``/mainprog/mainprog/__init__.py:``:

.. code-block:: python

    import polyvers

    __title__ = "mainprog"
    __version__ = polyvers.version('mainprog')
    ...

...and respectively ``/core-lib/core/__init__.py:``:

.. code-block:: python

    __version__ = polyvers.version('core')



Bump sub-projects selectively
-----------------------------
Now let's add another dummy commit and then bump ONLY ONE sub-project:

.. code-block:: console

    $ git commit  --allow-empty  -m "some head work"
    $ polyvers bump ^1 mainprog
    00:53:07       |NOTIC|polyvers.bumpcmd.BumpCmd|Bumped projects: mainprog-0.0.0 --> 0.0.1

    $ git lg
    COMMITS BRANCH TAGS                 REMARKS
    ======= ====== ==================== ========================================
         O  latest mainprog-r0.0.1.dev0 - The latest "Release" leaf-commit.
        /                                 branch `latest` was reset non-ff.
       O    MASTER mainprog-v0.0.1.dev0 - The latest "Version" commit.
       O                                - some head work
       | O         mainprog-r0.0.0      - Now it's obvious why "Release" commits
       |/          core-r0.0.0            are called "leafs".
       O           mainprog-v0.0.0
       |           core-v0.0.0
       O

    $ git checkout latest
    $ cat mainprog/mainprog/__init__.py
    import polyvers

    __title__     = "mainprog"
    __version__ = '0.0.1.dev0'
    ...

    $ cat core/core/__init__.py
    import polyvers

    __title__ = "core"
    __version__ = '0.0.0+mainprog.0.0.1.dev0'
    ...
    $ git checkout -

Notice how the the `"local" part of PEP-440
<https://www.python.org/dev/peps/pep-0440/#local-version-identifiers>`_ (statring with ``+...``)
is used by the engraved version of the **un-bumped** ``core`` project to signify
the correlated version of the **bumped** ``mainprog``.  This trick is not necessary
for tags because they apply repo-wide, to all sub-projects.


.. _features:

Features
========
.. include:: <xhtml1-lat1.txt>
.. glossary::

    PEP 440 version ids
        While most versioning tools use `Semantic versioning
        <http://semver.org/>`_, python's ``distutils`` native library
        supports the quasi-superset, but more versatile, `PEP-440 version ids
        <https://www.python.org/dev/peps/pep-0440/>`_, like that:

        - Pre-releases: when working on new features::

            X.YbN               # Beta release
            X.YrcN  or  X.YcN   # Release Candidate
            X.Y                 # Final release

        - Post-release::

            X.YaN.postM         # Post-release of an alpha release

CHANGES.rst
=======
Changes
=======

.. *towncrier* instructions for the developers:

    - Install cmd-line tool in your workstation::

            pip install towncrier

    - Add change-items in `changes.d/<issue>.<type>` files like that::

          echo 'Fixed a thing!' > ./changes.d/1234.fix

      Available change types:
        - break
        - feat
        - drop
        - fix
        - change
        - doc
        - chore

    - Generate new changes for the comming release::

          towncrier --draft       # to preview
          towncrier

    - *towncrier*'s config file is `pyproject.toml`.

. contents:: Releases
    :local:

.. _to-dos:

TODOs
=====
- Parse ``git describe`` like `setuptools_scm plugin
  <https://pypi.org/project/setuptools_scm/#default-versioning-scheme>`_ does.

- Drop `pvcmd/pvtags.py`, and replace it with `polyversion`?

- Engravings:  Not easy to extend!

  - Configurable hooks - refactor :term:`engravings` as one of them.
    to run, for example, housekeeping commands on all subprojects like
    ``pip install -e <project>`` and immediately start working in "develop mode".

    This would allow housekeeping commands and *validate tests*
    before/after every bump::

        ## Pre-release hook
        #
        pytest tests


        ## Post-release hook
        #
        rm -r dist/* build/*;
        python setup.py sdist bdist_wheel
        twine upload dist/*whl -s

  - Add top-level engrave glob-excludes.

  - Use `astor <https://pypi.org/project/astor/>`_ grafter.

- Refactor :term:`version-bump algebra` to support a single modifier per segment
  (see ``multivermath`` branch).

- Lock release-trains as "alpha/beta".., specific branches can be selected
  Based on :term:`version-bump algebra`), this will force users to always
  use ``pip install --pre`` to fetch such release-trains.
  This is a safeguard to avoid accidentally landing half-baked code to users.

- Retrofit `polyversion` library as a plugin of `polyvers` command.

- Function as plugin for other 3rd-party projects, bake a cookiecutter

- Check what happens with `no-commit`, e.g. to temporarily package a wheel.

.. towncrier release notes start


2018-08-07: polyversion-v0.2.2a1, polyvers-v0.1.1a1
===================================================
Maintenance release.

- enh:  `polyversion` now logs which was cmd not found on Windows
  (usually it is executing ``git``).
- chore: Merge all *pyupd* requests for dependencies.


2018-07-08: polyversion-v0.2.2a0
================================
- FIX: `git < 2.15.0` was buggy with multiple match-patterns in command::

      git describe --match=...  --match=...

  (see https://github.com/git/git/blob/master/Documentation/RelNotes/2.15.0.txt)

  The ``polyvers`` cmd still affected by this.


2018-07-06: polyvers-v0.1.1a0
=============================
FIX: engravings were applied unsorted (based on their start-point) and offsets
were miss-located when switching graft in a same file.


2018-07-05: polyversion-v0.2.1a0
================================
- Feature: The :envvar:`POLYVERSION_LOG_LEVEL` control *polyversion* verbosity.
  Run ``polyversion -h`` fo help.
- Change: minor reordering when searching version from package-metadata.
- fix: add standalone ``bin/pvlib.run`` from last release.
- fix: :func:`~polyversion.polyversion()`/:func:`~polyversion.polytime()`
  are guessing ``basepath`` keyword from the path of caller's top-package
  (not just from caller's fpath).


2018-07-04: polyversion-v0.2.0a2
================================
- Version `v0.2.0a0` not in pypi, consumed for standalone ``bin/pvlib.run``.
- Version `v0.2.0a1` were not finding sbling-dir versions if ``pip install git+...``,
  and had not replaced all skip-bdist flags.


Features
--------
- Teach non-engraved projects how to retrieve polyversion when pip-installed:

  - The functions :func:`~polyversion.polyversion()` & :func:`~polyversion.polytime()`
    now attempt to fetch version from package/site-package infos.
  - And the function doing this  :func:`polyversion.pkg_metadata_version()`
    retrofitted to:

    - search  for `<pname-<version>.egg-info/PKG-INFO` in `baspath` sibling folder
      (before searching PKG-INFO, METADATA in `basepath`),
    - so now `basepath` always needed in  ``polyversion()/polytime()`` functions
      to locate sibling dir.


Breaking Changes
----------------
- Rename :term:`setuptools` flag from ``skip_polyversion_check -->
  polyversion_check_bdist_enabled`` to flip its default logic (not checking by
  default), since non-engraved wheels install just fine now.
- Rename the keyword of ``polyversion()``/``polytime()`` functions from
  ``repo_path --> basepath`` to denote its importance for retrieving the version
  of installed projects from sibling dirs inside  ``PYTHONPATH/site-packages/``.


2018-06-29: polyversion-v0.1.1a3
================================
(change actually done in `v0.1.1a1`, just a fixes & doc-msg in `a2`)

- FIX: teach :term:`setuptools plugin` about :term:`default version env-var`.
  Now can ``pip install git+https://some.git.repo/but-from/non-engraved-branch``.


2018-06-27: polyversion-v0.1.1a0
================================
- FEAT: Introduce configurable :term:`default version env-var` to fall-back
  to :envvar:`<pname>_VERSION` if it exists, in case of errors (e.g. no git).
  The presence of such a variable also sets ``polytime(no_raise=True)``,
  which now also support the ``pname`` and ``default_version_env_var`` kwds.


2018-06-06: polyvers-v0.1.0a1, polyversion-v0.1.0a7
===================================================
Mostly docs, combined release.

+ FEAT: reinstated :term:`engravings` on ``_version.py``
  (see previous release for rational).


2018-06-05: polyvers-v0.1.0a0, polyversion-v0.1.0a6: co2mpas-ready
==================================================================
+ FEAT: reinstated :term:`engravings` on ``setup.py`` (dropped only for a while

polyvers/cli.py
# -*- coding: utf-8 -*-
#
# Copyright 2015-2018 European Commission (JRC);
# Licensed under the EUPL 1.2+ (the 'Licence');
# You may not use this work except in compliance with the Licence.
# You may obtain a copy of the Licence at: http://ec.europa.eu/idabc/eupl
#
"""The code of *polyvers* shell-commands."""

from collections import OrderedDict, defaultdict, Mapping
from pathlib import Path
from typing import Dict
from typing import Tuple, Set, List  # noqa: F401 @UnusedImport, flake8 blind in funcs
import io
import logging

from boltons.setutils import IndexedSet as iset

from . import APPNAME, __version__, __updated__, pvtags, pvproject, \
    polyversion as pvlib
from ._vendor import traitlets as trt
from ._vendor.traitlets import config as trc
from ._vendor.traitlets.traitlets import Bool, Unicode
from ._vendor.traitlets.traitlets import List as ListTrait, Tuple as TupleTrait
from .cmdlet import cmdlets, autotrait


log = logging.getLogger(__name__)


####################
## Config sources ##
####################
CONFIG_VAR_NAME = '%s_CONFIG_PATHS' % APPNAME.upper()
#######################


#: YAML dumper used to serialize command's outputs.
_Y = None


def _get_yamel():
    global _Y

    if not _Y:
        from ruamel import yaml
        from ruamel.yaml.representer import RoundTripRepresenter

        for d in [OrderedDict, defaultdict]:
            RoundTripRepresenter.add_representer(
                d, RoundTripRepresenter.represent_dict)
        _Y = yaml.YAML()

    return _Y


def ydumps(obj):
    "Dump any false objects as empty string, None as nothing, or as YAML. "

    if obj is None:
        return
    if not obj:
        return ''

    sio = io.StringIO()
    _get_yamel().dump(obj, sio)
    return sio.getvalue().strip()


def yloads(text):
    "Dump any false objects as empty string, None as nothing, or as YAML. "

    if not text:
        return

    return _get_yamel().load(text)


def merge_dict(dct, merge_dct):
    """ Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param dct: dict onto which the merge is executed
    :param merge_dct: dct merged into dct
    :return: None

    Adapted from: https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
    """
    for k in merge_dct.keys():
        if (k in dct and isinstance(dct[k], dict) and isinstance(merge_dct[k], Mapping)):
            merge_dict(dct[k], merge_dct[k])
        else:
            dct[k] = merge_dct[k]


class PolyversCmd(cmdlets.Cmd):
    """
    Bump independently PEP-440 versions of sub-project in Git monorepos.

    SYNTAX:
      {cmd_chain} <sub-cmd> ...
    """
    version = __version__
    examples = Unicode("""
        - Let it guess the configurations for your monorepo::
              {cmd_chain} init
          You may specify different configurations paths with::
              {cmd_chain} --config-paths /foo/bar/:~/{config_basename}.yaml:.

        - Use then the main sub-commands::
              {cmd_chain} init    # (optional) use it once, or to update configs.
              {cmd_chain} status
              {cmd_chain} bump 0.0.1.dev0 -c '1st commit, untagged'
              {cmd_chain} bump -t 'Mostly model changes, tagged'
    """)
    classes = [pvproject.Project]

    projects = ListTrait(
        autotrait.AutoInstance(pvproject.Project),
        config=True)

    @trt.default('subcommands')
    def _subcommands(self):
        subcmds = OrderedDict()
        subcmds['bump'] = ('polyvers.bumpcmd.BumpCmd',
                           "Increase or set (sub-)project version(s).")
        subcmds.update(cmdlets.build_sub_cmds(InitCmd, StatusCmd, LogconfCmd))
        subcmds['config'] = (
            'polyvers.cmdlet.cfgcmd.ConfigCmd',
            "Commands to inspect configurations and other cli infos.")

        return subcmds

    def collect_app_infos(self):
        """Provide extra infos to `config infos` subcommand."""
        return {
            'version': __version__,
            'updated': __updated__,
            ## TODO: add more app-infos.
        }

    @trt.default('all_app_configurables')
    def _all_app_configurables(self):
        from . import bumpcmd
        return [type(self),
                pvproject.Project,
                InitCmd, StatusCmd, bumpcmd.BumpCmd, LogconfCmd,
                pvproject.Engrave, pvproject.Graft,
                ]

    @trt.default('config_paths')
    def _config_paths(self):
        basename = self.config_basename
        paths = []

        git_root = fu.find_git_root()
        if git_root:
            paths.append(str(git_root / basename))
        else:
            paths.append('.')

        paths.append('~/%s' % basename)

        return paths

    _git_root: Path = None

    @property
    def git_root(self) -> Path:
        if self._git_root is None:
            self._git_root = fu.find_git_root()

            if not self._git_root:
                raise pvtags.NoGitRepoError(
                    "Current-dir '%s' is not inside a git-repo!" %
                    Path().resolve())

        return self._git_root

    autodiscover_subproject_projects = ListTrait(
        autotrait.AutoInstance(pvproject.Project),
        default_value=[{
            'engraves': [{
                'globs': ['**/setup.py'],
                'grafts': [
                    {'regex': r'''(?xm)
                        \b(name|PROJECT|APPNAME|APPLICATION)
                        \ *=\ *
                        (['"])
                            (?P<pname>[\w\-.]+)
                        \2
                    '''}
                ]
            }]
        }],
        allow_none=True,
        config=True,
        help="""
        Two projects with glob-patterns/regexes autodiscovering sub-project basepath/names.

        - Needed when a) no configuration file is given (or has partial infos),
          and b) when constructing/updating the configuration file.
        - The glob-patterns contained in this `Project[Engrave[Graft]]`
          should match files in the root dir of auto-discovered-projects
          (`Graft.subst` are unused here).
        - `Project.basepath` must be a relative
        - Regex(es) may extract the project-name from globbed files.
          If none (or different ones) match, project detection fails.
        - A Project is identified only if file(s) are globbed AND regexp(s) matched.
        """)

    autodiscover_version_scheme_projects = TupleTrait(
        autotrait.AutoInstance(pvproject.Project), autotrait.AutoInstance(pvproject.Project),
        default_value=(
            pvtags.make_match_all_pvtags_project(),
            pvtags.make_match_all_vtags_project(),
        ),
        config=True,
        help="""
        A pair of Projects with patterns/regexps matching *pvtags* or *vtags*, respectively.

        - Needed when a) no configuration file is given (or has partial infos),
          and b) when constructing/updating the configuration file.
        - Screams if discovered same project-name with conflicting basepaths.
        """)

    def _autodiscover_project_basepaths(self) -> Dict[str, Path]:
        """
        Invoked when no config exists (or asked to updated it) to guess projects.

        :return:
            a mapping of {pnames: basepaths}
        """
        from . import engrave

        if not self.autodiscover_subproject_projects:
            raise cmdlets.CmdException(
                "No `Polyvers.autodiscover_subproject_projects` param given!")

        fproc = engrave.FileProcessor(parent=self)
        with self.errlogged(doing='discovering project paths',
                            info_log=self.log.info):
            scan_projects = self.autodiscover_subproject_projects
            #: Dict[Path,
            #: List[Tuple[pvproject.Project, Engrave, Graft, List[Match]]]]
            match_map = fproc.scan_projects(scan_projects)

        ## Accept projects only if one, and only one,
        #  pair (pname <--> path) matched.
        #
        pname_path_pairs: List[Tuple[str, Path]] = [
            (m.groupdict()['pname'], fpath.parent)
            for fpath, mqruples in match_map.items()
            for _prj, _eng, _graft, matches in mqruples
            for m in matches]
        unique_pname_paths = iset(pname_path_pairs)

        ## check basepath conflicts.
        #
        projects: Dict[str, Path] = {}
        dupe_projects: Dict[str, Set[Path]] = defaultdict(set)
        for pname, basepath in unique_pname_paths:
            dupe_basepath = projects.get(pname)
            if dupe_basepath and dupe_basepath != basepath:
                dupe_projects[pname].add(basepath)
            else:
                projects[pname] = basepath

        if dupe_projects:
            raise cmdlets.CmdException(
                "Discovered conflicting project-basepaths: %s" %
                ydumps(dupe_basepath))

        return projects

    def _autodiscover_versioning_scheme(self):
        """
        Guess whether *monorepo* or *mono-project* versioning schema applies.

        :return:
            one of :func:`pvtags.make_vtag_project`, :func:`pvtags.make_pvtag_project`
        """
        pvtag_proj, vtag_proj = self.autodiscover_version_scheme_projects
        pvtags.populate_pvtags_history(pvtag_proj, vtag_proj)

        if bool(pvtag_proj.pvtags_history) ^ bool(vtag_proj.pvtags_history):
            return (pvtags.make_pvtag_project(parent=self)
                    if pvtag_proj.pvtags_history else
                    pvtags.make_vtag_project(parent=self))
        else:
            raise cmdlets.CmdException(
                "Cannot auto-discover versioning scheme, "
                "missing or contradictive versioning-tags:\n%s"
                "\n\n  Try --monorepo/--mono-project flags." %
                ydumps({'pvtags': pvtag_proj.pvtags_history,
                        'vtags': vtag_proj.pvtags_history}))

    def bootstrapp_projects(self) -> None:
        """
        Ensure valid configuration exist for monorepo/mono-project(s).

        :raise CmdException:
            if cwd not inside a git repo
        """
        git_root = self.git_root

        template_project = pvproject.Project(parent=self)
        has_template_project = (template_project.tag_vprefixes and
                                template_project.pvtag_frmt and
                                template_project.pvtag_regex)

        if not has_template_project:
            template_project = self._autodiscover_versioning_scheme()
            self.log.info("Auto-discovered versioning scheme: %s",
                          template_project.pname)

        has_subprojects = bool(self.projects)
        if not has_subprojects:
            proj_paths: Dict[str, Path] = self._autodiscover_project_basepaths()
            if not proj_paths:
                raise cmdlets.CmdException(
                    "Cannot auto-discover (sub-)project path(s)!"
                    "\n  Please use `ìnit` cmd to specify sub-projects explicitly.")

            self.log.info(
                "Auto-discovered %i sub-project(s) in git-root '%s': \n%s",
                len(proj_paths), git_root.resolve(),
                ydumps({k: str(v) for k, v in proj_paths.items()}))

            self.projects = [template_project.replace(pname=name,
                                                      basepath=basepath,
                                                      _pvtags_collected=None)
                             for name, basepath in proj_paths.items()]

        return self.projects


class _SubCmd(PolyversCmd):
    def __init__(self, *args, **kw):
        self.subcommands = {}
        super().__init__(*args, **kw)


class InitCmd(_SubCmd):
    """Generate configurations based on directory contents."""

    def _find_config_file_path(self, rootapp) -> Path:
        """
        Log if no config-file has been loaded.
        """
        git_root = rootapp.git_root

        for p in self._cfgfiles_registry.collected_paths:
            p = Path(p)
            try:
                if p.relative_to(git_root):
                    return p
            except ValueError as _:
                ## Ignored, probably program defaults.
                pass

    def run(self, *args):
        if len(args) > 0:
            raise cmdlets.CmdException(
                "Cmd %r takes no arguments, received %d: %r!"
                % (self.name, len(args), args))

        self.bootstrapp_projects()
        cfgpath = self._find_config_file_path(self)
        if cfgpath:
            yield "TODO: update config-file '%s'...." % cfgpath
        else:
            cfgpath = Path(self.git_root) / ('%s.yaml' % self.config_basename)
            yield "TODO: create new config-file in '%s'." % cfgpath


_status_all_help = """
    When true, fetch also all version-tags, otherwise just project version-id(s).
"""


class StatusCmd(_SubCmd):
    """
    List the versions of project(s).

    SYNTAX:
        {cmd_chain} [OPTIONS] [<project>]...
    """
    all = Bool(  # noqa: A003
        config=True,
        help=_status_all_help)

    flags = {'all': ({'StatusCmd': {'all': True}}, _status_all_help)}

    def _describe_projects(self, projects):
        def git_desc_without_screams(proj):
            try:
                return proj.git_describe()
            except pvtags.GitVoidError as _:
                return None

        versions = {p.pname: {'version': git_desc_without_screams(p)}
                    for p in projects}
        return versions

    def _fetch_all(self, projects):
        ## TODO: extract method to classify pre-populated histories.
        pvtags.populate_pvtags_history(*projects)
        ## TODO: YAMLable Project (apart from Printable) with metadata Print/header
        pinfos = {p.pname: {'history': p.pvtags_history,
                            'basepath': str(p.basepath)}
                  for p in projects}
        return pinfos

    def run(self, *pnames):
        projects = self.bootstrapp_projects()

        if pnames:
            ## TODO: use _filter_projects_by_name()
            projects = [p for p in projects
                        if p.pname in pnames]

        res = self._describe_projects(projects)

        if self.all:
            merge_dict(res, self._fetch_all(projects))

        if res:
            return ydumps(res)


class LogconfCmd(_SubCmd):
    """Write a logging-configuration file that can filter logs selectively."""
    def run(self, *args):
        pass


# TODO: Will work when patched: https://github.com/ipython/traitlets/pull/449
PolyversCmd.config_paths.tag(envvar=CONFIG_VAR_NAME)
trc.Application.raise_config_file_errors.tag(config=True)
trc.Application.raise_config_file_errors.help = \
    'Whether failing to load config files should prevent startup.'

PolyversCmd.flags = {  # type: ignore
    ## Copied from Application
    #
    'show-config': ({
        'Application': {
            'show_config': True,
        },
    }, trc.Application.show_config.help),
    'show-config-json': ({
        'Application': {
            'show_config_json': True,
        },
    }, trc.Application.show_config_json.help),

    ## Consulted by main.init_logging() if in sys.argv.
    #
    ('v', 'verbose'): (
        {'Spec': {'verbose': True}},
        cmdlets.Spec.verbose.help
    ),
    ('n', 'dry-run'): (
        {'Spec': {'dry_run': True}},
        cmdlets.Spec.dry_run.help
    ),
    ('d', 'debug'): ({
        'Spec': {
            'debug': True,
        }, 'Application': {
            'show_config': True,
            'raise_config_file_errors': True,
        }},
        cmdlets.Spec.debug.help
    ),

    ('a', 'amend'): (
        {'pvproject.Project': {'amend': True}},
        pvproject.Project.amend.help
    ),
    ('t', 'tag'): (
        {'Project': {'tag': True}},
        pvproject.Project.tag.help
    ),

    'monorepo': (
        {'Project': {
            'pvtag_frmt': pvlib.pvtag_frmt,
            'pvtag_regex': pvlib.pvtag_regex,
        }},
        "Use *pvtags* for versioning sub-projects in this git monorepo."
    ),
    'mono-project': (
        {'Project': {
            'pvtag_frmt': pvlib.vtag_frmt,
            'pvtag_regex': pvlib.vtag_regex,
        }},
        "Use plain *vtags* for versioning a single project in this git repo."
    ),
}

PolyversCmd.aliases = {  # type: ignore
    ('f', 'force'): 'Spec.force',
}


cmdlets.Spec.force.help += """
Supported tokens:
  'fread'     : don't stop engraving on file-reading errors.
  'fwrite'    : don't stop engraving on file-writting errors.
  'foverwrite': overwrite existing file.
  'glob'      : keep-going even if glob-patterns are invalid.
  'tag'       : replace existing tag.
"""
